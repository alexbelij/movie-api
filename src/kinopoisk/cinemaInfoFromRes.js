/* @flow */

import moment from 'moment';

import type {
  KinopoiskApi$GetCinemaDetailView,
  KinopoiskApi$MovieShowtimes,
  KinopoiskApi$Showtime,
} from './types';

const timestampForShowtime = ({
  time,
  date,
  utcOffset,
}: {
  time: string,
  date: string,
  utcOffset: string,
}) => {
  const timeMoment = moment(time, 'HH.mm');
  const dateMoment = moment(date, 'DD.MM.YYYY').utcOffset(utcOffset);

  if (!dateMoment) return null;

  const hours = timeMoment.hours();
  const minutes = timeMoment.minutes();

  return dateMoment
    .clone()
    .hours(hours < 4 ? hours + 24 : hours)
    .minutes(minutes)
    .toDate();
};

const showtimesFromRes = (
  {
    date,
    items,
  }: {
    date: string,
    items: Array<KinopoiskApi$MovieShowtimes>,
  } = {},
  utcOffset: ?string,
) => {
  if (!date || !items) return [];

  return items.map(({filmID, seance}: KinopoiskApi$MovieShowtimes) => ({
    filmId: parseInt(filmID, 10),
    items: seance.map(({time, formats}: KinopoiskApi$Showtime) => ({
      time,
      date,
      timestamp: utcOffset
        ? timestampForShowtime({time, date, utcOffset})
        : null,
      formats: formats.format,
    })),
  }));
};

const cinemaInfoFromRes = (
  {cinemaDetail}: KinopoiskApi$GetCinemaDetailView,
  utcOffset: ?string,
) =>
  !cinemaDetail
    ? null
    : {
        id: parseInt(cinemaDetail.cinemaID, 10),
        name: cinemaDetail.cinemaName,
        address: (cinemaDetail.cinemaLocation || {}).addressDescription,
        location: {
          lat: parseFloat((cinemaDetail.cinemaLocation || {}).lat),
          lng: parseFloat((cinemaDetail.cinemaLocation || {}).lon),
        },
        phoneNumbers: cinemaDetail.phones,
        homepage: cinemaDetail.cinemaWeb,
        showtimes: showtimesFromRes(cinemaDetail.seance, utcOffset),
      };

export {
  timestampForShowtime as __timestampForShowtime,
  showtimesFromRes as __showtimesFromRes,
};
export default cinemaInfoFromRes;
