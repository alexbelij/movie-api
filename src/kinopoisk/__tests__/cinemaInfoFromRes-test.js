/* @flow */

import { __timestampForShowtime } from '../cinemaInfoFromRes';

describe('kinopoisk/cinemaInfoFromRes', () => {
  it('generates a timestamp for showtime', () => {
    expect(__timestampForShowtime({
      time: '20:00',
      date: '14.11.2016',
      utcOffset: '+0600',
    })).toMatchSnapshot();

    expect(__timestampForShowtime({
      time: '02:00',
      date: '14.11.2016',
      utcOffset: '+0600',
    })).toMatchSnapshot();

    expect(__timestampForShowtime({
      time: '02:00',
      date: '14.11.2016',
      utcOffset: '-0600',
    })).toMatchSnapshot();
  });
});
