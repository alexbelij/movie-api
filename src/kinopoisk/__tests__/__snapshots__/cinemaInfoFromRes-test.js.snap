exports[`kinopoisk/cinemaInfoFromRes generates a timestamp for showtime 1`] = `2016-11-14T14:00:00.000Z`;

exports[`kinopoisk/cinemaInfoFromRes generates a timestamp for showtime 2`] = `2016-11-14T20:00:00.000Z`;

exports[`kinopoisk/cinemaInfoFromRes generates a timestamp for showtime 3`] = `2016-11-14T08:00:00.000Z`;
